### PR TITLE
Per-tenant elasticsearch configuration

### PR DESCRIPTION
**Pull request by caseydavenport**
## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/logstorage/secrets/secret_controller.go [hunk 1]
```diff
@@ -495,6 +500,14 @@ func (r *SecretSubController) collectUpstreamCerts(log logr.Logger, helper utils
 		certificatemanagement.CASecretName: common.OperatorNamespace(),
 	}
 
+	if r.elasticExternal {
+		// For external ES, we don't need to generate a keypair for ES itself. Instead, a public certificate
+		// for the external ES and Kibana instances must be provided. Load and include in these into
+		// the trusted bundle for Linseed and es-gateway.
+		certs["tigera-secure-es-http-certs-public"] = common.OperatorNamespace()
```
Comment (caseydavenport) [inline pkg/controller/logstorage/secrets/secret_controller.go:507]: Should make constants for these.

File: pkg/controller/manager/manager_controller.go [hunk 2]
```diff
@@ -167,8 +167,10 @@ func Add(mgr manager.Manager, opts options.AddOptions) error {
 		return fmt.Errorf("manager-controller failed to watch the '%s' namespace: %w", common.TigeraPrometheusNamespace, err)
 	}
 
-	if err = utils.AddConfigMapWatch(managerController, render.ECKLicenseConfigMapName, render.ECKOperatorNamespace, eventHandler); err != nil {
-		return fmt.Errorf("manager-controller failed to watch the ConfigMap resource: %v", err)
+	if !opts.ElasticExternal {
```
Comment (caseydavenport) [inline pkg/controller/manager/manager_controller.go:170]: I wonder if this watch is actually needed at all? I don't see the controller ever reference this configmap.

File: pkg/render/common/elasticsearch/service.go [hunk 3]
```diff
@@ -49,5 +50,12 @@ func HTTPSEndpoint(osType rmeta.OSType, clusterDomain string) string {
 }
 
 func ElasticEndpoint() string {
+	if ep := os.Getenv("ELASTIC_ENDPOINT"); ep != "" {
```
Comment (caseydavenport) [inline pkg/render/common/elasticsearch/service.go:53]: This is making me rethink putting the elastic URL on the Tenant resource. We probably need a per-cluster fallback anyway, which sort of invalidates that concept. 

Unless I can think of a way to make this work, we'll probably need to switch back to a per-cluster ES URL.

File: main.go [hunk 4]
```diff
@@ -68,6 +71,10 @@ var (
 	setupLog                 = ctrl.Log.WithName("setup")
 )
 
+// bootstrapConfigMapName is the name of the ConfigMap that contains cluster-wide
+// configuration for the operator loaded at startup.
+const bootstrapConfigMapName = "operator-bootstrap-config"
```
Comment (tmjd) [inline main.go:76]: Should we include any notes or warnings that this ConfigMap is not a stable API and should be avoided, or maybe that it is only intended for unsupported enterprise features.

File: pkg/controller/logstorage/elastic/external_elastic_controller.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	}
+	if err = imageset.AddImageSetWatch(c); err != nil {
+		return fmt.Errorf("log-storage-external-es-controller failed to watch ImageSet: %w", err)
+	}
+	if err = c.Watch(&source.Kind{Type: &operatorv1.ManagementCluster{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("log-storage-external-es-controller failed to watch ManagementCluster resource: %w", err)
+	}
+	if err = c.Watch(&source.Kind{Type: &operatorv1.ManagementClusterConnection{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("log-storage-external-es-controller failed to watch ManagementClusterConnection resource: %w", err)
+	}
```
Comment (tmjd) [inline pkg/controller/logstorage/elastic/external_elastic_controller.go:92]: I don't see these being used by this controller, am I just missing it?

File: pkg/controller/logstorage/elastic/external_elastic_controller.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	if err = utils.AddTigeraStatusWatch(c, tigeraStatusName); err != nil {
+		return fmt.Errorf("log-storage-external-es-controller failed to watch logstorage Tigerastatus: %w", err)
+	}
+
+	if err = utils.AddConfigMapWatch(c, "cloud-kibana-config", common.OperatorNamespace(), &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("log-storage-external-es-controller failed to watch the ConfigMap resource: %w", err)
+	}
```
Comment (tmjd) [inline pkg/controller/logstorage/elastic/external_elastic_controller.go:99]: I don't see this being used either.

File: pkg/controller/logstorage/elastic/external_elastic_controller.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+	if err = imageset.AddImageSetWatch(c); err != nil {
+		return fmt.Errorf("log-storage-external-es-controller failed to watch ImageSet: %w", err)
+	}
+	if err = c.Watch(&source.Kind{Type: &operatorv1.ManagementCluster{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("log-storage-external-es-controller failed to watch ManagementCluster resource: %w", err)
+	}
+	if err = c.Watch(&source.Kind{Type: &operatorv1.ManagementClusterConnection{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("log-storage-external-es-controller failed to watch ManagementClusterConnection resource: %w", err)
+	}
+	if err = utils.AddTigeraStatusWatch(c, tigeraStatusName); err != nil {
```
Comment (tmjd) [inline pkg/controller/logstorage/elastic/external_elastic_controller.go:93]: We don't need this do we for statusManager do we?

**Review by tmjd — COMMENTED** [review]
What is the plan/expectation for the kube-controllers image used with external ES being a different build?

File: pkg/controller/manager/manager_controller.go (hunk 2 above)
Comment (tmjd) [reply pkg/controller/manager/manager_controller.go:170]: :shrug:

**Comment by caseydavenport** [comment]
@tmjd fixed!

**Review by tmjd — COMMENTED** [review]
LGTM, leaving for @asincu 's approval too. (I might have to come back and give a final approval since I don't know if you have enough permissions to meet the PR requirements.)

File: pkg/controller/logstorage/linseed/linseed_controller.go [hunk 8]
```diff
@@ -374,19 +409,22 @@ func (r *LinseedSubController) Reconcile(ctx context.Context, request reconcile.
...
+		KeyPair:             linseedKeyPair,
+		TokenKeyPair:        tokenKeyPair,
+		UsePSP:              r.usePSP,
+		ESClusterConfig:     esClusterConfig,
+		HasDPIResource:      hasDPIResource,
+		ManagementCluster:   managementCluster != nil,
+		Tenant:              tenant,
+		ElasticHost:         elasticHost,
+		ElasticPort:         elasticPort,
+		ElasticClientSecret: esClientSecret,
```
Comment (asincu) [inline pkg/controller/logstorage/linseed/linseed_controller.go:427]: If we are using mTLS, this secret also contains the CA, right ? `ElasticClientSecret`makes me think we only store the client certificate.

File: pkg/render/logstorage/linseed/linseed.go [hunk 9]
```diff
@@ -299,6 +311,28 @@ func (l *linseed) linseedDeployment() *appsv1.Deployment {
...
+			},
+		})
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      logstorage.ExternalCertsVolumeName,
+			MountPath: "/certs/elasticsearch/mtls",
+			ReadOnly:  true,
+		})
+
+		// Configure Linseed to use the mounted client certificate and key.
+		envVars = append(envVars, corev1.EnvVar{Name: "ELASTIC_MTLS_ENABLED", Value: "true"})
```
Comment (asincu) [inline pkg/render/logstorage/linseed/linseed.go:331]: Don;t we also need to override the ca path also?

File: pkg/render/logstorage/linseed/linseed_test.go [hunk 10]
```diff
@@ -124,6 +127,47 @@ var _ = Describe("Linseed rendering tests", func() {
...
+			}))
+			Expect(d.Spec.Template.Spec.Containers[0].VolumeMounts).To(ContainElement(corev1.VolumeMount{
+				Name:      logstorage.ExternalCertsVolumeName,
+				MountPath: "/certs/elasticsearch/mtls",
+				ReadOnly:  true,
+			}))
+
+			// The client secret should also be emitted as a resources, but copied to the destination namespace.
+			s, ok := rtest.GetResource(createResources, logstorage.ExternalCertsSecret, render.ElasticsearchNamespace, "", "v1", "Secret").(*corev1.Secret)
+			Expect(ok).To(BeTrue(), "Secret not copied")
```
Comment (asincu) [inline pkg/render/logstorage/linseed/linseed_test.go:167]: Should we add some checks for the environment variables that are set when mTLS is enabled ?

File: api/v1/tenant_types.go [hunk 11]
```diff
@@ -66,8 +70,14 @@ type TenantSpec struct {
...
 	ID string `json:"id,omitempty"`
 
+	// Name is a human readable name for this tenant.
+	Name string `json:"name,omitempty"`
+
 	// Indices defines the how to store a tenant's data
 	Indices []Index `json:"indices"`
+
+	// Elastic configures per-tenant ElasticSearch and Kibana parameters.
+	Elastic *TenantElasticSpec `json:"elastic,omitempty"`
```
Comment (asincu) [inline api/v1/tenant_types.go:80]: When can this field be missing ?

File: main.go (hunk 4 above)
Comment (asincu) [reply main.go:77]: I assume we would also want to have this configured by helm when configuring a multi-tenant management cluster.

File: pkg/controller/utils/utils.go [hunk 12]
```diff
@@ -779,3 +783,54 @@ func GetDNSServiceName(provider operatorv1.Provider) types.NamespacedName {
...
+			cs.CoreV1().RESTClient(),
+			"configmaps",
+			common.OperatorNamespace(),
+			fields.OneTermEqualSelector("metadata.name", name),
+		),
+		&v1.ConfigMap{},
+		0, // no resync period
+	)
+	_, err := informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, newObj interface{}) {
```
Comment (asincu) [inline pkg/controller/utils/utils.go:800]: What happens if the configmap is deleted ?

File: pkg/render/common/elasticsearch/service.go (hunk 3 above)
Comment (asincu) [reply pkg/render/common/elasticsearch/service.go:53]: Should we add a task in the backlog ?

File: pkg/render/logstorage/external.go [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+const (
+	// Secret and volume name used for client certificate and key. Used by Linseed and es-gateway
+	// when mTLS to external Elasticsearch is enabled. The secret contains the client certificate
+	// and key to present to elastic.
+	ExternalCertsSecret     = "tigera-secure-external-es-certs"
+	ExternalCertsVolumeName = "tigera-secure-external-es-certs"
+
+	// ExternalESPublicCertName and ExternalKBPublicCertName are the names of the public certificates
+	// used as part of CA buncldes to trust external Elasticsearch and Kibana instances.
```
Comment (asincu) [inline pkg/render/logstorage/external.go:25]: typo: bundles

File: pkg/render/logstorage/external.go [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logstorage
+
+const (
+	// Secret and volume name used for client certificate and key. Used by Linseed and es-gateway
+	// when mTLS to external Elasticsearch is enabled. The secret contains the client certificate
+	// and key to present to elastic.
+	ExternalCertsSecret     = "tigera-secure-external-es-certs"
```
Comment (asincu) [inline pkg/render/logstorage/external.go:21]: Does it also contain the CA ?

File: pkg/controller/manager/manager_controller.go [hunk 15]
```diff
@@ -166,11 +166,6 @@ func Add(mgr manager.Manager, opts options.AddOptions) error {
 	if err = utils.AddNamespaceWatch(managerController, common.TigeraPrometheusNamespace); err != nil {
 		return fmt.Errorf("manager-controller failed to watch the '%s' namespace: %w", common.TigeraPrometheusNamespace, err)
 	}
-
-	if err = utils.AddConfigMapWatch(managerController, render.ECKLicenseConfigMapName, render.ECKOperatorNamespace, eventHandler); err != nil {
```
Comment (asincu) [inline pkg/controller/manager/manager_controller.go:170]: :thinking: Why were we watching the license here in the first place ?

File: pkg/controller/logstorage/linseed/linseed_controller.go (hunk 8 above)
Comment (caseydavenport) [reply pkg/controller/logstorage/linseed/linseed_controller.go:427]: The CA is not within this secret. The CA is contained within the trusted bundle that's already provided, so you're correct that this only contains the client certificate and key. 

The comment on the field explains this also: https://github.com/tigera/operator/blob/0ec02e4228f794c340c5f9f7440129575d9aa70c/pkg/render/logstorage/linseed/linseed.go#L120-L122

File: pkg/render/logstorage/linseed/linseed.go (hunk 9 above)
Comment (caseydavenport) [reply pkg/render/logstorage/linseed/linseed.go:331]: Nope, the CA remains the same.

File: api/v1/tenant_types.go (hunk 11 above)
Comment (caseydavenport) [reply api/v1/tenant_types.go:80]: This field can be omitted if not using an external elasticsearch. I'll add a comment

File: pkg/controller/utils/utils.go (hunk 12 above)
Comment (caseydavenport) [reply pkg/controller/utils/utils.go:800]: Good question - I copied this code largely from existing CC code, which didn't have a deletion handler. I think in practice deleting this configmap will break the cluster, so it sort of doesn't matter what we do (and if it's deleted / re-created we'll get an ADD event which will be handled)

File: pkg/render/logstorage/external.go (hunk 14 above)
Comment (caseydavenport) [reply pkg/render/logstorage/external.go:21]: nope

File: pkg/controller/manager/manager_controller.go (hunk 15 above)
Comment (caseydavenport) [reply pkg/controller/manager/manager_controller.go:170]: No clue, that's why I removed it :sweat_smile:

File: pkg/render/logstorage/linseed/linseed_test.go (hunk 10 above)
Comment (asincu) [reply pkg/render/logstorage/linseed/linseed_test.go:175]: I think we also need an `ELASTIC_MTLS_ENABLED` check ....

**Review by asincu — APPROVED** [review]
With a small comment

**Pull request merged**